llvm/test/CodeGen/AArch64/vector-promotion.ll

```diff
@@ -0,0 +1,434 @@
+; ASM-NEXT:    ldp x20, x19, [sp], #32 ; 16-byte Folded Reload
+; ASM-NEXT:    ret
+  %in1 = load <2 x float>, <2 x float>* %addr1, align 8
+  %extract = extractelement <2 x float> %in1, i32 1
+  %out = frem nnan float 7.0, %extract
+  store float %out, float* %dest, align 4
+  ret void
+}
+
+; Check that we are able to promote floating point value.
+; Scalar version:
+; Vector version:
+;
+define void @simpleOneInstructionPromotionFloat(<2 x float>* %addr1, float* %dest) {
+; ASM-LABEL: simpleOneInstructionPromotionFloat:
+; ASM:       ; %bb.0:
+; ASM-NEXT:    ldr d0, [x0]
+; ASM-NEXT:    fmov s1, #1.00000000
+; ASM-NEXT:    mov s0, v0[1]
+; ASM-NEXT:    fadd s0, s0, s1
+; ASM-NEXT:    str s0, [x1]
+; ASM-NEXT:    ret
+  %in1 = load <2 x float>, <2 x float>* %addr1, align 8
+  %extract = extractelement <2 x float> %in1, i32 1
+  %out = fadd float %extract, 1.0
+  store float %out, float* %dest, align 4
+  ret void
+}
+
+; Check that we correctly use a splat constant when we cannot
+; determine at compile time the index of the extract.
+; This requires the STRESS modes, as variable index are expensive
+; to lower.
+; Scalar version:
+; Vector version:
+;
+define void @simpleOneInstructionPromotionVariableIdx(<2 x i32>* %addr1, i32* %dest, i32 %idx) {
+; ASM-LABEL: simpleOneInstructionPromotionVariableIdx:
+; ASM:       ; %bb.0:
+; ASM-NEXT:    ; kill: def $w2 killed $w2 def $x2
+; ASM-NEXT:    and x8, x2, #0x1
+; ASM-NEXT:    ldr w8, [x0, x8, lsl #2]
+; ASM-NEXT:    orr w8, w8, #0x1
+; ASM-NEXT:    str w8, [x1]
+; ASM-NEXT:    ret
+  %in1 = load <2 x i32>, <2 x i32>* %addr1, align 8
+  %extract = extractelement <2 x i32> %in1, i32 %idx
+  %out = or i32 %extract, 1
+  store i32 %out, i32* %dest, align 4
+  ret void
+}
+
+; Check a vector with more than 2 elements.
+; This requires the STRESS mode because currently 'or v8i8' is not marked
+; as legal or custom, althought the actual assembly is better if we were
+; promoting it.
+; Scalar version:
+; Vector version:
+;
+define void @simpleOneInstructionPromotion8x8(<8 x i8>* %addr1, i8* %dest) {
+; ASM-LABEL: simpleOneInstructionPromotion8x8:
+; ASM:       ; %bb.0:
+; ASM-NEXT:    movi.8b v0, #1
+; ASM-NEXT:    ldr d1, [x0]
+; ASM-NEXT:    orr.8b v0, v1, v0
+; ASM-NEXT:    st1.b { v0 }[1], [x1]
+; ASM-NEXT:    ret
+  %in1 = load <8 x i8>, <8 x i8>* %addr1, align 8
+  %extract = extractelement <8 x i8> %in1, i32 1
+  %out = or i8 %extract, 1
+  store i8 %out, i8* %dest, align 4
+  ret void
+}
+
+; Check that we optimized the sequence correctly when it can be
+; lowered on a Q register.
+;
+; Make sure we got rid of any expensive vmov.32 instructions.
+; The Q register used here must be [[LOAD]] / 2, but we cannot express that.
+define void @simpleOneInstructionPromotion4x32(<4 x i32>* %addr1, i32* %dest) {
+; ASM-LABEL: simpleOneInstructionPromotion4x32:
+; ASM:       ; %bb.0:
+; ASM-NEXT:    ldr q0, [x0]
+; ASM-NEXT:    orr.4s v0, #1
+; ASM-NEXT:    st1.s { v0 }[1], [x1]
+; ASM-NEXT:    ret
```

Comment:
Currently, this generates:
```
        ldr     w8, [x0, #4]
        orr     w8, w8, #0x1
        str     w8, [x1]
        ret
```
Which seems like it could be preferred over vector operations.  Similar for `simpleOneInstructionPromotion8x8`.